src/v/archival/ntp_archiver_service.cc

```diff
@@ -915,9 +917,82 @@ ss::future<cloud_storage::upload_result> ntp_archiver::upload_tx(
       get_bucket_name(), manifest, fib, _tx_tags);
 }
 
+ss::future<std::optional<cloud_storage::offset_index>>
+ntp_archiver::make_segment_index(
+  const upload_candidate& candidate,
+  retry_chain_logger& ctxlog,
+  std::string_view index_path) {
+    auto base_rp_offset = candidate.starting_offset;
+    auto base_kafka_offset = model::offset_cast(
+      _parent.get_offset_translator_state()->from_log_offset(base_rp_offset));
+
+    cloud_storage::offset_index ix{
+      base_rp_offset,
+      base_kafka_offset,
+      0,
+      cloud_storage::remote_segment_sampling_step_bytes};
+
+    vlog(ctxlog.debug, "creating remote segment index: {}", index_path);
+    auto builder = cloud_storage::make_remote_segment_index_builder(
+      storage::concat_segment_reader_view(
+        candidate.sources,
+        candidate.file_offset,
+        candidate.final_file_offset,
+        _conf->upload_io_priority)
+        .take_stream(),
```

Comment:
This will work but you're essentially reading the index twice, first to upload the segment and then to generate the index. This could be very noticeable for our clients. I think it's better to read the index once. You can do this by first creating a stream and then splitting it into two streams using `input_stream_fanout` utility from `stream_utils.h`.